### ISSUE TITLE
Mac ARM64 build fails when cross compiling from a Mac Intel

### ISSUE THREAD

**Issue by rofferom** [getsentry/sentry-native#445]
## Description

I'm trying to build sentry for Mac ARM64 (cross compiling from a Mac Intel), but I got some build issues. I managed to workaround them, but I don't know how to do make a correct fix (my cmake knowledge is really bad). So I'm going to describe the issue and explain what I found.

Note that I'm using a standalone Xcode install, placed at `/Applications/Xcode12.3.app`.

I'm working with SHA-1 026fc6441322f2f3133bb3eee798ff2fdf73af63

## Build setup

Standard cmake + make build :
```
$ mkdir build && cd build
$ export DEVELOPER_DIR=/Applications/Xcode12.3.app/Contents/Developer/
$ cmake -DCMAKE_OSX_DEPLOYMENT_TARGET=11.0 -DCMAKE_OSX_ARCHITECTURES=arm64 -DCMAKE_OSX_SYSROOT=/Applications/Xcode12.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk ..
make -j4
```

## Issue 1

The build fails when calling `external/crashpad/util/mach/mig.py`:
```
[  9%] Generating util/mach/child_portUser.c, util/mach/child_portServer.c, util/mach/child_port.h, util/mach/child_portServer.h
cd /tmp/sentry-native/build/crashpad_build/util && /usr/bin/python2.7 /tmp/sentry-native/external/crashpad/util/mach/mig.py --arch=arm64 --include=/tmp/sentry-native/external/crashpad/util/.. --include=/tmp/sentry-native/external/crashpad/util/../compat/mac /tmp/sentry-native/external/crashpad/util/mach/child_port.defs /tmp/sentry-native/build/crashpad_build/util/util/mach/child_portUser.c /tmp/sentry-native/build/crashpad_build/util/util/mach/child_portServer.c /tmp/sentry-native/build/crashpad_build/util/util/mach/child_port.h /tmp/sentry-native/build/crashpad_build/util/util/mach/child_portServer.h
In file included from /tmp/sentry-native/external/crashpad/util/mach/child_port.defs:15:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/mach/mach_types.defs:168:
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/mach/machine/thread_state.h:35:2: error: architecture not supported
#error architecture not supported
 ^
1 error generated.
"/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/mach/mach_types.defs", line 170: syntax error
mig: fatal: "/tmp/sentry-native/external/crashpad/util/mach/child_port.defs", line 62: 1 errors found. Abort.
```

Actually it seems that it is not using the SDK I want:
```
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/mach/machine/thread_state.h:35:2: error: architecture not supported
```

`external/crashpad/util/mach/mig.py` can get a `--sdk` option. I have made a dirty change to add `--sdk`, and the error is now fixed:
```
diff --git a/util/CMakeLists.txt b/util/CMakeLists.txt
index 7a3425f8..f77763b1 100644
--- a/util/CMakeLists.txt
+++ b/util/CMakeLists.txt
@@ -387,7 +387,7 @@ if(APPLE)
             OUTPUT
             ${output_files}
             COMMAND
-            "${PYTHON_EXECUTABLE}" "${CMAKE_CURRENT_SOURCE_DIR}/mach/mig.py" ${archs} ${includes} "${input}" ${output_files}
+            "${PYTHON_EXECUTABLE}" "${CMAKE_CURRENT_SOURCE_DIR}/mach/mig.py" ${archs} --sdk ${CMAKE_OSX_SYSROOT} ${includes} "${input}" ${output_files}
             DEPENDS
             "${CMAKE_CURRENT_SOURCE_DIR}/mach/mig.py" "${input}"
         )
```

## Error 2

I get another error during `libcrashpad_util.a` link:
```
[ 39%] Linking CXX static library libcrashpad_util.a
/Applications/Xcode12.3.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: archive member: libcrashpad_util.a(capture_context_mac.S.o) cputype (16777223) does not match previous archive members cputype (16777228) (all members must match)
/Applications/Xcode12.3.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib libcrashpad_util.a
/Applications/Xcode12.3.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: archive member: libcrashpad_util.a(capture_context_mac.S.o) cputype (16777223) does not match previous archive members cputype (16777228) (all members must match)
make[2]: *** [crashpad_build/util/libcrashpad_util.a] Error 1
make[2]: *** Deleting file `crashpad_build/util/libcrashpad_util.a'
make[1]: *** [crashpad_build/util/CMakeFiles/crashpad_util.dir/all] Error 2
make: *** [all] Error 2
```

The file `capture_context_mac.S.o` has been built for x86_64:
```
$ file ./crashpad_build/util/CMakeFiles/crashpad_util.dir/misc/capture_context_mac.S.o
./crashpad_build/util/CMakeFiles/crashpad_util.dir/misc/capture_context_mac.S.o: Mach-O 64-bit object x86_64
```

To be sure, I have checked that a random `.o` file is correctly built for arm64:
```
$ file ./crashpad_build/util/CMakeFiles/crashpad_util.dir/misc/time.cc.o 
./crashpad_build/util/CMakeFiles/crashpad_util.dir/misc/time.cc.o: Mach-O 64-bit object arm64
```

Let's check the compilation flags.
```
$ touch ../external/crashpad/util/misc/capture_context_mac.S
$ make
[ 10%] Building ASM object crashpad_build/util/CMakeFiles/crashpad_util.dir/misc/capture_context_mac.S.o
cd /tmp/sentry-native/build/crashpad_build/util && /Applications/Xcode12.3.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -DCRASHPAD_LSS_SOURCE_EMBEDDED -DCRASHPAD_ZLIB_SOURCE_SYSTEM -DZLIB_CONST -I/tmp/sentry-native/build/crashpad_build/util -I/tmp/sentry-native/external/crashpad -I/tmp/sentry-native/external/crashpad/third_party/mini_chromium/mini_chromium -I/tmp/sentry-native/external/crashpad/compat/non_win -I/tmp/sentry-native/external/crashpad/compat/mac -I/tmp/sentry-native/external/crashpad/compat/non_elf -I/Applications/Xcode12.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -O3 -g -DNDEBUG -isysroot /Applications/Xcode12.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -mmacosx-version-min=11.0 -fPIC -o CMakeFiles/crashpad_util.dir/misc/capture_context_mac.S.o -c /tmp/sentry-native/external/crashpad/util/misc/capture_context_mac.S
```

It seems that no `-arch` parameter isn't passed. Another dirty change to add the missing flag fixes the build error:
```
diff --git a/util/CMakeLists.txt b/util/CMakeLists.txt
index 7a3425f8..f77763b1 100644
@@ -410,6 +410,10 @@ target_link_libraries(crashpad_util
 )
 
 if(APPLE)
+    get_property(archs TARGET crashpad_util PROPERTY OSX_ARCHITECTURES)
+    list(TRANSFORM archs PREPEND "-arch ")
+    set(CMAKE_ASM_FLAGS "${CFLAGS} ${archs}")
+
     target_link_libraries(crashpad_util PRIVATE
         bsm
         "-framework CoreFoundation"
```

### Build success :ok_hand: 

```
$ file libsentry.dylib ./crashpad_build/handler/crashpad_handler
libsentry.dylib:                           Mach-O 64-bit dynamically linked shared library arm64
./crashpad_build/handler/crashpad_handler: Mach-O 64-bit executable arm64

```

**Comment by Swatinem** [comment]
We maintain our own crashpad fork here: https://github.com/getsentry/crashpad

I would very much appreciate a PR, since you basically have a patch ready already ;-)

**Comment by rofferom** [comment]
Done

**Comment by Swatinem** [comment]
Thanks so much! Next step is to bump the submodule here in sentry-native

**Comment by rofferom** [comment]
It seems that the fix has been integrated in version 0.4.5 (#452) :ok_hand:

**Comment by Swatinem** [comment]
Yup ;-)
